app/cli/commandlineparser.cpp

```diff
@@ -0,0 +1,363 @@
+#include "commandlineparser.h"
+
+#include <QCommandLineParser>
+#include <QRegularExpression>
+
+#if defined(Q_OS_WIN)
+#include <qt_windows.h>
+#endif
+
+static bool inRange(int value, int min, int max)
+{
+    return value >= min && value <= max;
+}
+
+// This method returns key's value from QMap where the key is a QString.
+// Key matching is case insensitive.
+template <typename T>
+static T mapValue(QMap<QString, T> map, QString key)
+{
+    for(auto& item : map.toStdMap()) {
+        if (QString::compare(item.first, key, Qt::CaseInsensitive) == 0) {
+            return item.second;
+        }
+    }
+    return T();
+}
+
+class CommandLineParser : public QCommandLineParser
+{
+public:
+    enum MessageType {
+        Info,
+        Error
+    };
+
+    void setupCommonOptions()
+    {
+        setSingleDashWordOptionMode(QCommandLineParser::ParseAsLongOptions);
+        addHelpOption();
+        addVersionOption();
+    }
+
+    void handleHelpAndVersionOptions()
+    {
+        if (isSet("help")) {
+            showInfo(helpText());
+        }
+        if (isSet("version")) {
+            showVersion();
+        }
+    }
+
+    void handleUnknownOptions()
+    {
+        if (unknownOptionNames().length()) {
+            showError(QString("Unknown options: %1").arg(unknownOptionNames().join(", ")));
+        }
+    }
+
+    void showMessage(QString message, MessageType type) const
+    {
+    #if defined(Q_OS_WIN)
+        UINT flags = MB_OK | MB_TOPMOST | MB_SETFOREGROUND;
+        flags |= (type == Info ? MB_ICONINFORMATION : MB_ICONERROR);
+        QString title = "Moonlight";
+        MessageBoxW(nullptr, reinterpret_cast<const wchar_t *>(message.utf16()),
+                    reinterpret_cast<const wchar_t *>(title.utf16()), flags);
+    #endif
+        message = message.endsWith('\n') ? message : message + '\n';
+        fputs(qPrintable(message), type == Info ? stdout : stderr);
+    }
+
+    [[ noreturn ]] void showInfo(QString message) const
+    {
+        showMessage(message, Info);
+        exit(0);
+    }
+
+    [[ noreturn ]] void showError(QString message) const
+    {
+        showMessage(message, Error);
```

Comment:
Can we include the help text in this message? Maybe like `message + "\n\n" + helpText()`?

Reply:
Can do